File: lib/utils/heksagon_djeyometre.dart
import 'dart:math';
import '../models/angol_modalz.dart';

class HeksagonDjeyometre {
  final double hexSize;
  final HexagonPosition center;
  final bool isLetterMode;

  HeksagonDjeyometre({
    required this.hexSize,
    required this.center,
    this.isLetterMode = true,
  });

  double get hexWidth => sqrt(3) * hexSize;
  double get hexHeight => 2 * hexSize;
  double get rotationAngle => 0.0;

  HexagonPosition axialToPixel(int q, int r) {
    double x, y;
    x = hexSize * (sqrt(3) * q + sqrt(3) / 2 * r);
    y = hexSize * (3 / 2) * r;
    return HexagonPosition(
      x: center.x + x,
      y: center.y + y,
    );
  }

  List<AxialCoordinate> getInnerRingCoordinates() {
    return const [
      AxialCoordinate(q: 1, r: -1),
      AxialCoordinate(q: 1, r: 0),
      AxialCoordinate(q: 0, r: 1),
      AxialCoordinate(q: -1, r: 1),
      AxialCoordinate(q: -1, r: 0),
      AxialCoordinate(q: 0, r: -1),
    ];
  }

  List<AxialCoordinate> getOuterRingCoordinates() {
    return const [
      AxialCoordinate(q: 2, r: -2),
      AxialCoordinate(q: 2, r: -1),
      AxialCoordinate(q: 2, r: 0),
      AxialCoordinate(q: 1, r: 1),
      AxialCoordinate(q: 0, r: 2),
      AxialCoordinate(q: -1, r: 2),
      AxialCoordinate(q: -2, r: 2),
      AxialCoordinate(q: -2, r: 1),
      AxialCoordinate(q: -2, r: 0),
      AxialCoordinate(q: -1, r: -1),
      AxialCoordinate(q: 0, r: -2),
      AxialCoordinate(q: 1, r: -2),
    ];
  }
}

File: kepad/src/commonMain/kotlin/yuteledez/HeksagonDjeyometre.kt
package yuteledez

import modalz.AxialCoordinate
import modalz.HexagonPosition
import kotlin.math.sqrt

class HeksagonDjeyometre(
    val hexSize: Double,
    val center: HexagonPosition,
    val isLetterMode: Boolean = true
) {
    val hexWidth: Double
        get() = sqrt(3.0) * hexSize

    val hexHeight: Double
        get() = 2 * hexSize

    val rotationAngle: Double = 0.0

    /**
     * Converts axial coordinates (q, r) to pixel coordinates (x, y).
     */
    fun axialToPixel(q: Int, r: Int): HexagonPosition {
        // In Kotlin, integer division `3 / 2` would result in 1.
        // We use floating-point numbers to match the original Dart behavior.
        val x = hexSize * (sqrt(3.0) * q + sqrt(3.0) / 2.0 * r)
        val y = hexSize * (1.5) * r // 3.0 / 2.0 = 1.5
        return HexagonPosition(
            x = center.x + x,
            y = center.y + y
        )
    }

    /**
     * Returns the list of axial coordinates for the inner ring of hexagons.
     */
    fun getInnerRingCoordinates(): List<AxialCoordinate> {
        return listOf(
            AxialCoordinate(q = 1, r = -1),
            AxialCoordinate(q = 1, r = 0),
            AxialCoordinate(q = 0, r = 1),
            AxialCoordinate(q = -1, r = 1),
            AxialCoordinate(q = -1, r = 0),
            AxialCoordinate(q = 0, r = -1)
        )
    }

    /**
     * Returns the list of axial coordinates for the outer ring of hexagons.
     * Note: The original Dart code had 12 coordinates here.
     */
    fun getOuterRingCoordinates(): List<AxialCoordinate> {
        return listOf(
            AxialCoordinate(q = 2, r = -2),
            AxialCoordinate(q = 2, r = -1),
            AxialCoordinate(q = 2, r = 0),
            AxialCoordinate(q = 1, r = 1),
            AxialCoordinate(q = 0, r = 2),
            AxialCoordinate(q = -1, r = 2),
            AxialCoordinate(q = -2, r = 2),
            AxialCoordinate(q = -2, r = 1),
            AxialCoordinate(q = -2, r = 0),
            AxialCoordinate(q = -1, r = -1),
            AxialCoordinate(q = 0, r = -2),
            AxialCoordinate(q = 1, r = -2)
        )
    }
}
